plugins/modules/vm_powerstate.py

```
@@ -0,0 +1,468 @@
+        """
+        if self.module.params['folder']:
+            self.module.params['folder'] = format_folder_path_as_vm_fq_path(
+                self.module.params['folder'],
+                self.module.params['datacenter']
+            )
+
+    def run_vm_configuration_task(self, task):
+        if task:
+            try:
+                _, task_result = RunningTaskMonitor(task).wait_for_completion(
+                    vm=self.vm, timeout=self.module.params['timeout'], answers=self.module.params.get('question_answers', None))
+            except TaskError as e:
+                self.module.fail_json(msg=to_text(e))
+            finally:
+                    self.result['changed'] = True
+
+    def reset_vm(self):
+        if self.current_state not in ('poweredon', 'poweringon', 'resetting', 'poweredoff'):
+            self.module.fail_json(msg="Cannot restart virtual machine in the current state %s" % self.current_state)
+        return self.vm.Reset()
+
+    def suspend_vm(self):
+        if self.current_state not in ('poweredon', 'poweringon'):
+            self.module.fail_json(msg='Cannot suspend virtual machine in the current state %s' % self.current_state)
+        return self.vm.Suspend()
+
+    def set_vm_state_using_vmtools(self):
+        if self.current_state == 'poweredon':
+            if self.vm.guest.toolsRunningStatus == 'guestToolsRunning':
+                if self.desired_state == 'shutdownguest':
+                    task = self.vm.ShutdownGuest()
+                else:
+                    task = self.vm.RebootGuest()
+                # Set result['changed'] immediately because
+                # shutdown and reboot return None.
+                self.result['changed'] = True
+            else:
+                self.module.fail_json(msg="VMware tools should be installed for guest shutdown/reboot")
+        elif self.current_state == 'poweredoff':
+            self.result['changed'] = False
+            return None
```

this seems odd to me. I think it should raise an error or something?

---

i guess it depends on desired state. if the user wanted to reboot then this should be an error, otherwise this looks right